### ISSUE TITLE
Messages larger than MAX_CONTIGUOUS_MESSAGE_DATA (9000 bytes) do not work

### ISSUE THREAD

**Issue by jmw182** [dmw109/pyrtma#7]
Messages larger than 9000 bytes (e.g. TIMING_MESSAGE) were handled by the C++ message manager by using the `char * large_data` property of the CMessage class to point to the data, and by sending the header and data separately.

Right now receiving a large message sent from a module causes an error in `read_message()` as it tries to assign the incoming data to `msg.data,` which is too small. We also need to send `TIMING_MESSAGE` from message manager, which is currently not possible, since we cannot assign the large data to `msg.data`

My first thought for a solution is to create a `LargeMessage` class which accepts a `num_data_bytes` value in the constructor and uses this instead of `MAX_CONTIGUOUS_MESSAGE_DATA` to set the data length. A `LargeTimeCodeMessage` would also be needed.

Or, it might be possible to continue using `DefaultMessage` and `TimeCodeMessage`, but add `num_data_bytes` as an optional input, which defaults to `MAX_CONTIGUOUS_MESSAGE_DATA`

**Comment by dmw109** [comment]
Are we actually using these large messages? I think we should only replicate if it is used by us.

**Comment by jmw182** [comment]
At the very least, the TIMING_MESSAGE which contains module pids (and is used by the BCI loader) is affected by this. I'm not sure if any of our other message types are > 9000 bytes.

**Comment by dmw109** [comment]
I am not sure why the first field has to be so crazy large.  What is this message for?

**Comment by dmw109** [comment]
Maybe we should make Message a regular class that is composed of two ctype structs, the header and data.  Then we could choose data to be whatever size the header indicates.

Maybe it could have a method that sets the correct data type after the header is read.

You could also return a dynamically created class def from a function.  Though I am not sure how that works on multiple calls to the function.

**Comment by jmw182** [comment]
> I am not sure why the first field has to be so crazy large. We never have 10000 messages going at once. What is this message for?

That's true, if we want to change the message definition we could replace the large `timing` field with two shorter fields, one with active message IDs and a second with the message counts. Ideally it would function more like a python dictionary (we just care about pairs of message IDs and message counts for those IDs). The only issue is those vectors would still need to be long enough to contain all message types that are currently in use, so we would still need to define a constant like `MAX_MESSAGE_TYPES_PER_SESSION` and decide what that value should be.

TIMING_MESSAGE is used for two purposes in the bci Loader: 
1. the `ModulePID` field has the list of process IDs for each active module (message manager maintains this list by collecting pids when a module calls `SendModuleReady()`). Loader uses this to know if a module is active or not. This state is displayed by a checkbox for each module, and clicking the button in Loader will either send the command to launch or close a module depending on whether it is currently active or not. Some modules need to be closed using the pid instead of an `EXIT` signal (by calling a taskkill command).
2. Loader displays message rates (in Hz) for a subset of message types (the subset is defined in Loader). Rather than subscribe to each message of interest and compute this timing individually for each message, message manager counts how many messages of each type are sent between sending each `TIMING_MESSAGE`, and Loader just needs to use those counts and divide by the time interval between when each `TIMING_MESSAGE` is sent.

**Comment by dmw109** [comment]
Do we have to recompile to change that constant? If so, then I think we just support the larger size, but not promote its use.

We may be able to build something dynamically. 
```
msg_cls_types ={}
def new_message(header_type, msg_size):
    try:
         return msg_cls_types[(header_type, msg_size)]

    except KeyError:
        class Message(ctypes.Structure):
            _fields_ = [('header', header_type), ('data', ctypes.c_byte * msg_size)]

          msg_cls_types[(header_type, msg_size)] = Message
          return Message
```

**Comment by jmw182** [comment]
I like that idea. We can also use `MAX_CONTIGUOUS_MESSAGE_DATA` as a default value if `msg_size` is not specified when constructing the class.

Which constant are you asking about? `MAX_CONTIGUOUS_MESSAGE_DATA`? And what do you mean be "recompile?" Rebuilding message defs? Rebuilding c++ rtma/message manager?

**Comment by dmw109** [comment]
I was referring to adding the max message types per session constant idea. I was thinking that would require recompiling the old message manager.

**Comment by jmw182** [comment]
Yes, any change to TIMING_MESSAGE would require recompiling the old message manager, unless we replaced it with a new message ID (and updated Loader to subscribe to both IDs and use either).

**Comment by dmw109** [comment]
If we always call recv twice for reading messages (once for the header and once for the data). Then maybe it would be ok to call send twice in send_message.  Then we can make Message a regular class composed of fields of two ctypes structs, header and data.  If we call send twice then it is not important that header and data are in a continuous memory buffer if that makes sense.

**Comment by jmw182** [comment]
Yeah, that makes sense to me.

**Comment by dmw109** [comment]
The only benefit of having it the way it is now is that we can call send once which may not matter anyways.
       



`

**Comment by jmw182** [comment]
Commit d22618a fixes this, at least from the manager.py side. I added `new_large_msg_cls` static methods to the the Message, DefaultMessage, and TimeCodeMessage classes to make a large message class with an arbitrary number of bytes. I updated manager.py to send the TIMING_MESSAGE and also tested that it can forward this this type sent from MATLAB.

I still need to test if anything needs to change to read and send large message types from PyRTMA clients.
